src/go/configmanager/config_manager_test.go

```diff
@@ -1800,37 +466,36 @@ func runTest(t *testing.T, opts options.ConfigGeneratorOptions, f func(*testEnv)
 	if err != nil {
 		t.Fatal("fail to initialize Config Manager: ", err)
 	}
-	env := &testEnv{
-		configManager: manager,
-	}
-	f(env)
+
+	f(manager)
 }
 
-func initMockConfigServer(t *testing.T) *httptest.Server {
-	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		_, err := w.Write(fakeConfig)
-		if err != nil {
-			t.Fatal("fail to write config: ", err)
-		}
-	}))
+type safeData struct {
+	mutex sync.Mutex
+	data  []byte
 }
 
-func initMockRolloutServer(t *testing.T) *httptest.Server {
-	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		w.Header().Set("Content-Type", "application/json")
-		_, err := w.Write(fakeRollouts)
-		if err != nil {
-			t.Fatal("fail to write rollout config: ", err)
-		}
-	}))
+func (s *safeData) write(src []byte) {
+	s.mutex.Lock()
```

Comment:
I think this will not fix the problem.  For each test, its always write() and read(), they are always sequential. 
The key problem is: multiple tests share the same `fakeConfig`,   each has different config data, and they are running in parallel so their data over-write each over.
The right approach is that each test should use its own object to store its config data for initMockServer.  We don't need to protect it,  as I see it,  we always create an object first (write) and use it (read) later.  As long as each test is using its own storage, they will not conflict with each other.

Reply:
I don't think so. If you open the sponge of test history, 
- only `TestFixedModeDynamicRouting` are failed and other two tests are passed
- the log only a sequential execution 
so it is not because of interference between tests. 

This is because the goroutine created by http server to handle request is reading the data and main goroutine is writing the data. This is an [example](https://stackoverflow.com/questions/31792810/simple-race-condition-in-go-http-handler-is-this-really-a-race-condition).

Reply:
I though for each test,  the main routine write the data, and pass it to httpServer, its handler only read the data.  read data can be in multiple threads.   which main routine is writing again? if not from different tests.

Reply:
That is exactly the case.

Reply:
in that case, let us not use a global variable for all test cases,  define one variable for each test case,  it should solve the problem.  not need to have mutex.

Reply:
Here, we are testing configmanager under `managed` mode so the config will be read by mockServer and modified by test in the same time.

Reply:
I see.  for multiple iterations.

Reply:
Thanks for the clarifications, I understand now